vmm/src/vm.rs

```diff
@@ -481,16 +483,30 @@ impl Vm {
         Ok(())
     }
 
-    pub fn resize(&mut self, desired_vcpus: u8) -> Result<()> {
-        self.cpu_manager
-            .lock()
-            .unwrap()
-            .resize(desired_vcpus)
-            .map_err(Error::CpuManager)?;
-        self.devices
-            .notify_hotplug(HotPlugNotificationType::CPUDevicesChanged)
-            .map_err(Error::DeviceManager)?;
-        self.config.lock().unwrap().cpus.boot_vcpus = desired_vcpus;
+    pub fn resize(&mut self, desired_vcpus: Option<u8>, desired_memory: Option<u64>) -> Result<()> {
+        if let Some(desired_vcpus) = desired_vcpus {
+            self.cpu_manager
+                .lock()
+                .unwrap()
+                .resize(desired_vcpus)
+                .map_err(Error::CpuManager)?;
+            self.devices
+                .notify_hotplug(HotPlugNotificationFlags::CPU_DEVICES_CHANGED)
+                .map_err(Error::DeviceManager)?;
+            self.config.lock().unwrap().cpus.boot_vcpus = desired_vcpus;
+        }
+
+        if let Some(desired_memory) = desired_memory {
+            self.memory_manager
+                .lock()
+                .unwrap()
+                .resize(desired_memory)
+                .map_err(Error::MemoryManager)?;
+            self.devices
+                .notify_hotplug(HotPlugNotificationFlags::MEMORY_DEVICES_CHANGED)
+                .map_err(Error::DeviceManager)?;
```

Comment:
I'm not asking to do that with this PR, but I think we're leaking memory and KVM slots if that fails as the guest may not ever see the new memory. So we should at least consider how to do downsizing when the notification fails. As we can assume the guest has never touched those pages, we could safely remove the DIMM.

Reply:
I think in the case of an error we should log something with error! and return the error to the user but not try and clean-up. This `.notify_hotplug()` method just generates an interrupt. If that's failing then I think we've got bigger problems.

Generally if the user can recover from the error then we should do some cleanup (or validate early) i.e. the requested size is invalid. In the cases where there is no clear path for the user to recover nor would retrying work then we should avoid the hard-to-test and potentially buggy attempt to cleanup.